changelog/4416.feature.rst

```diff
@@ -0,0 +1,6 @@
+pdb: support keyword arguments with ``pdb.set_trace``
+
+It handles ``header`` similar to Python 3.7 does it, and forwards any
+other keyword arguments to the ``Pdb`` constructor.
+
+This allows for ``__import__("pdb").set_trace(skip=["foo.*"])``.
```

Comment:
this changes the signature of `pdb.set_trace()` I don't think we should do this (we also have to manually adjust our code every time cpython adds a parameter to this function which seems non-ideal)

Reply:
the changelog text is incorrect, it should talk about `pytest.set_trace`

Reply:
`pytest` monkeypatches `pdb.set_trace`:

```
import pdb
import pytest


def test():
    assert pdb.set_trace.__module__ == '_pytest.debugging'
```

```console
$ pytest t.py -q
.                                                                        [100%]
1 passed in 0.00 seconds
```

Reply:
See https://github.com/pytest-dev/pytest/issues/4416#issuecomment-439710750.

We're not changing the signature, but just pass through args and kwargs.
`pdb.set_trace` looks like it might never handle args (`*, header=None` in py37), but just in case and/or if you want to hack something into this yourself.
This PR handles `header` added in py37, but in general (through the terminal writer), and forwards anything else to `Pdb.__init__`.
pdb++ makes use of this, and any custom pdbcls could do so then.

I do not see a compatibility issue really, but it makes it just a bit more forward-compatible - also a new kwarg (like "header") might cause an error then when `Pdb.__init__` does not handle it.

For this, we could try it with `**kwargs` first, and then without them on `TypeError`, logging a warning.

Reply:
@asottile right now, there is not a good way to provide those arguments to Pdb directly. If you could read into the original request #4416, I think it could help to give a better understanding on the context of the change.

Right now we are constrained to build the Pdb class in the predefined way, with no arguments. However with this patch we would be able to initalize Pdb, or any other implementation (I am keen on TerminalPdb from ipython, or pycharm's pdb class) with the custom arguments to support debugging asyncio.

I would highly encourage to try debugging asyncio code with `pytest.set_trace()` with no skip support, and you will find that all the control changes from the ioloop and back are almost impossible to follow due to all the await syntax.

This will ease a lot my development of fully asyncio libraries/applications, as many times you find bugs through pytest, and you need to debug them in that specific fixtured environment.

Reply:
JFI: as a workaround you could already use a custom `--pdbcls` setting/class I suppose, where you would handle/set `skip` in its `__init__` method.
Similar to https://github.com/pytest-dev/pytest/pull/4419/files#diff-6244362351d7c228d95cde366267b10aR659, but you would have to hardcode it there instead.

Reply:
my main concern is that by fundamentally changing what the arguments to `set_trace` do we're signing up for maintenance work as they change over time.  I get that it's inconvenient to properly set up the `Pdb` class in the way the cpython docs intend you to do it, but I don't think that's a strong enough reason to change the `set_trace` function to initialize `Pdb`.

Reply:
> we're signing up for maintenance work as they change over time

We have to do that already, as can be seen for handling the `header` kwarg here, which could maybe go into its own PR already.

And then I think it is better to forward unknown kwargs to `Pdb.__init__`  rather then not handling them altogether.

What do you think about trying it and logging a warning in case of `TypeError`?

Reply:
I'd rather we only handle the args intended for `set_trace` and `TypeError` on unknown arguments :man_shrugging: -- passing unknown arguments along to `Pdb.__init__` is what feels wrong to me

Reply:
I don't see what is the problem @asottile, if we pass the unknown arguments to `Pdb.__init__`, it will probably blow it up itself, no? 

What do you fear will happen in practice?

Reply:
my concern is that if we change how the `set_trace` function works only in a `pytest` context, it'll be mysteriously broken when you're not in pytest.

Consider the following:

```python
def f_under_test():
    import pdb; pdb.set_trace(skip=['django'])
    return 5

def test():
    assert f_under_test() == 5


def main():
    f_under_test()

if __name__ == '__main__':
    exit(main())
```

```
$ pytest t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.6, pytest-4.0.1.dev34+gd4a3f235.d20181122, py-1.7.0, pluggy-0.8.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/pytest/.hypothesis/examples')
rootdir: /tmp/pytest, inifile: tox.ini
plugins: hypothesis-3.82.1
collected 1 item                                                               

t.py 
>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>
> /tmp/pytest/t.py(3)f_under_test()
-> return 5
(Pdb) c

>>>>>>>>>>>>>>>>>>>>> PDB continue (IO-capturing resumed) >>>>>>>>>>>>>>>>>>>>>>
.                                                                   [100%]

=========================== 1 passed in 1.58 seconds ===========================

$ python t.py 
Traceback (most recent call last):
  File "t.py", line 13, in <module>
    exit(main())
  File "t.py", line 10, in main
    f_under_test()
  File "t.py", line 2, in f_under_test
    import pdb; pdb.set_trace(skip=['django'])
TypeError: set_trace() got an unexpected keyword argument 'skip'
```

Reply:
The root of it is monkeypatching is bad, and monkeypatching with changed behavior is worse

If it is permitted in pytest, someone will find it and assume that's how pdb works and get bit later. That pdbpp also commits this (imo mistake) (I mean, it's not the worst thing their packaging does :laughing:) should not be precedent

Reply:
Yeah, after all extending pdb in Python is not easy - and therefor pdbpp goes full-on with monkey-patching.  In a similar way to pytest btw, but it is more customizable.

After all, do we agree that you are not using pdb in production, but only only on demand, and then it's not really important if it fails early already?

Reply:
btw: are you using raw pdb really, or something else like pudb?

Reply:
Yeah that's fair, :shipit:. I actually do just use pdb, that way I don't have to install things

Reply:
Just in case, this is about a debugging line you would insert while developing tests.

I don't believe anyone would ever use `pytest.set_trace()` to develop in a normal day. It's specific to pytest. If you want to use pdb, you can always use pdb directly and instantiate it yourself.

I don't think we should reject the patch based on the argument that we have a function that is called the same and accepts more arguments...

Do you propose an alternative to feed arguments from pytest into pdb?

Reply:
@txomon the problem here is pytest monkepatches `pdb.set_trace = pytest.set_trace` -- if we were just changing `pytest.set_trace` I'd have no opposition to this patch

Reply:
I see @asottile's purist point of view, but I agree that `set_trace()` is really only used interactively so IMHO I don't believe this will break things or be surprising to users.

Reply:
Any news on this folks? @asottile do you still strongly object to this?

Reply:
Would help to get @RonnyPfannschmidt's opinion on this as well.

Reply:
I'd put myself at [non blocking opposed](https://github.com/pytest-dev/pytest/pull/4419#discussion_r235834258) go ahead!

Reply:
Heh, I can see @RonnyPfannschmidt literally on some fence already.. ;)

I'm also a bit on @asottile's side here in general, but do not really see a good way around this.  Maybe we could use `inspect` to have expected/proper TypeErrors?

Also, this removes the `break` kwarg that could previously have been used to do `pdb.set_trace(break=False)` - although it appears to come only from internal use without refactoring it properly (54d3cd58), i.e. an API break as it stands now.

Reply:
I think the thing that convinces me that this is alright is that an *interactive* call, so the user knows he is running its own debugger tracing function. If this was an API call I would also be against it, as you don't know what debugger the person using your API is.

But let's see @RonnyPfannschmidt's opinion, if he is also against this as @asottile is, then I'm afraid we will need to close this PR.